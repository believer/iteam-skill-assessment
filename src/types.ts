export enum MovieType {
  Game = 'game',
  Movie = 'movie',
  Series = 'movie',
}

export interface SearchMovie {
  Poster: string
  Title: string
  Type: MovieType
  Year: string
  imdbID: string
}

export interface SearchResults {
  Search: SearchMovie[]
}

export interface Rating {
  Source: string
  Value: string
}

export interface Movie {
  Actors: string
  Awards: string
  BoxOffice: string
  Country: string
  DVD: string
  Director: string
  Genre: string
  Language: string
  Metascore: string
  Plot: string
  Poster: string
  Production: string
  Rated: string
  Ratings: Rating[]
  Released: string
  Response: string
  Runtime: string
  Title: string
  Type: MovieType
  Website: string
  Writer: string
  Year: string
  imdbID: string
  imdbRating: string
  imdbVotes: string
}

export type MovieResponse = Movie & { Error: string }
