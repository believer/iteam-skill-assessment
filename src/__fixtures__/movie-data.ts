import { build, fake } from 'test-data-bot'
import { SearchMovie, MovieType, Movie } from '../types'

type PartialMovie = Pick<
  Movie,
  'Actors' | 'Poster' | 'Title' | 'Type' | 'Year' | 'imdbID'
>

export const searchMovieBuilder = build<SearchMovie>('Movie').fields({
  Poster: fake(f => f.internet.url()),
  Title: fake(f => f.lorem.words()),
  Type: MovieType.Movie,
  Year: fake(f =>
    f.date
      .past()
      .getFullYear()
      .toString()
  ),
  imdbID: fake(f => f.random.alphaNumeric()),
})

export const movieBuilder = build<PartialMovie>('Movie').fields({
  Actors: 'test, test2, test3',
  Poster: fake(f => f.internet.url()),
  Title: fake(f => f.lorem.words()),
  Type: MovieType.Movie,
  Year: fake(f =>
    f.date
      .past()
      .getFullYear()
      .toString()
  ),
  imdbID: fake(f => f.random.alphaNumeric()),
})
