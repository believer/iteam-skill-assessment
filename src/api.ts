import useFetch from 'react-fetch-hook'
import { MovieResponse, SearchResults } from './types'

const API_KEY = process.env.REACT_APP_OMDB_KEY

const createRequest = (params: string) =>
  `https://www.omdbapi.com/?apikey=${API_KEY}${params}`

export const useSearch = (query: string) => {
  const { data, isLoading, error } = useFetch<SearchResults>(
    createRequest(`&s=${query}`),
    {
      depends: [query],
    }
  )

  if (!data) {
    return { movies: [], isLoading, error }
  }

  return { movies: data.Search || [], isLoading, error }
}

export const useMovie = (id?: string) => {
  const { data, isLoading, error } = useFetch<MovieResponse>(
    createRequest(`&i=${id}`),
    {
      depends: [id],
    }
  )

  if (!data) {
    return { movie: null, isLoading, error }
  }

  if (data.Error) {
    return { movie: null, isLoading, error: data.Error }
  }

  return {
    movie: data,
    isLoading,
    error,
  }
}
