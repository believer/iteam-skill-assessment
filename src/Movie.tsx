import * as React from 'react'
import { RouteComponentProps, Link } from '@reach/router'
import { useMovie } from './api'
import { Grid, GridColumn } from './Grid'
import { H1, Paragraph } from './Typography'
import styled from '@emotion/styled'
import { Poster } from './MovieResult'
import { LoadingBlock } from './LoadingBlock'

interface BackdropProps {
  src: string
}

const Backdrop = styled.div<BackdropProps>`
  background-image: url(${({ src }) => src});
  background-size: cover;
  height: 300px;
  grid-column: -1/1;

  @media (min-width: 52em) {
    height: 500px;
  }
`

const Actor = styled.div`
  align-items: center;
  display: flex;
  padding: 5px;
`

const Avatar = styled.img`
  border-radius: 50%;
  height: 50px;
  margin-right: 15px;
  object-fit: cover;
  width: 50px;
`

const Name = styled.div`
  font-weight: 700;
`

const Role = styled.div`
  color: hsl(210, 12%, 60%);
  font-size: 12px;
  margin-top: 5px;
`

const BackLink = styled(Link)`
  color: #24292e;
  display: block;
  margin-bottom: 30px;
  margin-top: 30px;
`

const Movie: React.FC<RouteComponentProps<{ id: string }>> = ({ id }) => {
  const { movie } = useMovie(id)

  if (Object.keys(movie).length === 0) {
    return null
  }

  return (
    <Grid mb={100}>
      {movie.backdrop && <Backdrop src={movie.backdrop} />}
      <GridColumn>
        <BackLink to="/">Back to search</BackLink>
        <Grid mt={40} gridTemplateColumns={['1fr', '200px 1fr']}>
          <div>
            <Poster src={movie.poster} alt={movie.title} />
          </div>
          <div>
            <H1>{movie.title}</H1>
            <Paragraph mb="30px">{movie.overview}</Paragraph>
            <Grid
              gridRowGap="10px"
              gridTemplateColumns={['1fr', 'repeat(2, 1fr)']}
            >
              {movie.cast.slice(0, 15).map(actor => (
                <Actor key={actor.id}>
                  {!actor.image && (
                    <LoadingBlock
                      borderRadius="50%"
                      height="50px"
                      mr="15px"
                      width="50px"
                    />
                  )}
                  {actor.image && <Avatar src={actor.image} alt="" />}
                  <div>
                    <Name>{actor.name}</Name>
                    <Role>{actor.character}</Role>
                  </div>
                </Actor>
              ))}
            </Grid>
          </div>
        </Grid>
      </GridColumn>
    </Grid>
  )
}

export default Movie
