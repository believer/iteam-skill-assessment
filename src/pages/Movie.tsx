import styled from '@emotion/styled'
import { Link, RouteComponentProps } from '@reach/router'
import * as React from 'react'
import { useMovie } from '../api'
import { Grid, GridColumn } from '../components/Grid'
import { Poster } from '../components/MovieResult'
import { H1, Paragraph } from '../components/Typography'

const Actor = styled.div`
  align-items: center;
  display: flex;
  padding: 5px;
`

const Name = styled.div`
  font-weight: 700;
`

const Small = styled.div`
  color: hsl(210, 12%, 60%);
  font-size: 12px;
  margin-top: 5px;
`

const Meta = styled.div`
  display: flex;

  > *:not(:last-child) {
    margin-right: 20px;
  }
`

const BackLink = styled(Link)`
  color: #24292e;
  display: block;
  margin-bottom: 30px;
  margin-top: 30px;

  @media (min-width: 52em) {
    margin-top: 0;
  }
`

const ActorName = styled.div`
  flex: 1;
`

const PosterWrap = styled.div`
  justify-self: center;
`

const Movie: React.FC<RouteComponentProps<{ id: string }>> = ({ id }) => {
  const { movie } = useMovie(id)

  if (!movie) {
    return null
  }

  return (
    <Grid mb={100}>
      <GridColumn bg="#fff" borderRadius="10px" p={[0, 40]} mt={[0, -200]}>
        <BackLink to="/">Back to search</BackLink>
        <Grid mt={40} gridTemplateColumns={['1fr', '200px 1fr']}>
          <PosterWrap>
            <Poster src={movie.Poster} alt={movie.Title} />
          </PosterWrap>
          <div>
            <H1>{movie.Title}</H1>
            <Meta>
              <Small>{movie.Released}</Small>
              <Small>{movie.Genre}</Small>
            </Meta>
            <Paragraph mb="30px">{movie.Plot}</Paragraph>
            <Grid
              gridRowGap="10px"
              gridTemplateColumns={['1fr', 'repeat(2, 1fr)']}
            >
              {movie.Actors.split(', ').map(actor => (
                <Actor key={actor}>
                  <ActorName>
                    <Name>{actor}</Name>
                  </ActorName>
                </Actor>
              ))}
            </Grid>
          </div>
        </Grid>
      </GridColumn>
    </Grid>
  )
}

export default Movie
