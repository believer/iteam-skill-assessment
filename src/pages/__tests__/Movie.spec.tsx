import { render } from '@testing-library/react'
import React from 'react'
import { useMovie } from '../../api'
import { movieBuilder } from '../../__fixtures__/movie-data'
import Movie from '../Movie'
import { MovieType } from '../../types'

jest.mock('../../api')

test('renders loading state', () => {
  ;(useMovie as jest.Mock).mockReturnValue({
    movie: null,
    isLoading: true,
    error: null,
  })

  const { container } = render(<Movie id="1" />)

  expect(container).toMatchSnapshot()
})

test('renders error state', () => {
  ;(useMovie as jest.Mock).mockReturnValue({
    movie: null,
    isLoading: false,
    error: 'Incorrect IMDb ID',
  })

  const { container } = render(<Movie id="1" />)

  expect(container).toMatchSnapshot()
})

test('renders empty movie', () => {
  ;(useMovie as jest.Mock).mockReturnValue({
    movie: null,
    isLoading: false,
    error: null,
  })

  const { queryByText } = render(<Movie id="1" />)

  expect(useMovie).toHaveBeenCalledWith('1')

  // Tests movie title
  expect(queryByText(/batman/i)).not.toBeInTheDocument()
})

test('no runtime for games', () => {
  const movie = movieBuilder({ Type: MovieType.Game })
  ;(useMovie as jest.Mock).mockReturnValue({
    movie,
    isLoading: false,
    error: null,
  })

  const { queryByText } = render(<Movie id="1" />)

  expect(useMovie).toHaveBeenCalledWith('1')

  expect(queryByText(movie.Runtime)).not.toBeInTheDocument()
})

test('renders movie information', () => {
  const movie = movieBuilder()
  ;(useMovie as jest.Mock).mockReturnValue({
    movie,
    isLoading: false,
    error: null,
  })

  const { getByText } = render(<Movie id="1" />)

  expect(useMovie).toHaveBeenCalledWith('1')

  // Tests movie title
  expect(getByText(movie.Title)).toBeInTheDocument()

  // Tests cast
  expect(getByText(movie.Actors.split(', ')[0])).toBeInTheDocument()

  // Tests production company
  expect(getByText(movie.Production)).toBeInTheDocument()
})
