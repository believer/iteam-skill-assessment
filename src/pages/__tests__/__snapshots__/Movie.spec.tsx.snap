// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders error state 1`] = `
<div>
  <div
    class="h-screen flex items-center flex-col justify-center"
  >
    <div
      class="text-red-600 text-xl mb-2"
    >
      Incorrect IMDb ID
    </div>
    <a
      aria-current="page"
      class="text-gray-700"
      href="/"
    >
      Back to search
    </a>
  </div>
</div>
`;

exports[`renders loading state 1`] = `
<div>
  <div
    class="grid grid-template my-10"
  >
    <div
      class="grid-template-center"
    >
      <div
        class="grid grid-gap-12 grid-template-movie mt-8"
      >
        <div
          class="rounded bg-gray-200 w-full h-80"
        />
        <div>
          <div
            class="rounded bg-gray-200 w-2/12 h-8"
          />
          <div
            class="rounded bg-gray-200 w-6/12 h-5 mt-2"
          />
          <div
            class="rounded bg-gray-200 w-9/12 h-5 mt-5"
          />
          <div
            class="rounded bg-gray-200 w-9/12 h-5 mt-2"
          />
          <div
            class="grid grid-template-2 grid-gap-2 mt-8"
          >
            <div
              class="rounded bg-gray-200 w-full h-8"
            />
            <div
              class="rounded bg-gray-200 w-full h-8"
            />
            <div
              class="rounded bg-gray-200 w-full h-8"
            />
            <div
              class="rounded bg-gray-200 w-full h-8"
            />
          </div>
          <div
            class="mt-8"
          >
            <div
              class="rounded bg-gray-200 w-full h-8 mt-2"
            />
            <div
              class="rounded bg-gray-200 w-full h-8 mt-2"
            />
            <div
              class="rounded bg-gray-200 w-full h-8 mt-2"
            />
            <div
              class="rounded bg-gray-200 w-full h-8 mt-2"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
